src/corelibs/U2Core/src/globals/Nullable.h

```diff
@@ -45,10 +45,6 @@ class NP {
         return value != anotherValue;
     }
 
-    operator T*() const {
```

Comment:
This method should stay. Reason: all smart pointers have it.

You may think that using explicit .get() instead of auto-cast will help but this is not correct: you do not see the whole life-range of the pointer.
NP is designed to find errors in runtime (during GUI testing)

Reply:
I'll take it back, but I had SO many errors because of it, that this is the less friendly interface I've ever met

Reply:
I'm happy that this utility class helped you to find bugs: unchecked access to nullable pointers.
Except this I see no other non-friendly usage patterns: the NP pointer behaves just like a regular pointer with a 0 boilerplate.

Reply:
The only thing this class helped me is to create bugs and spend time. But OK, I'll take it back

Reply:
The main problem is that it's very difficult to track either you used `operator==` (or `operator!=`) or not before some line. You need to check, in some cases, more than several hundred lines to to find out if some `NL<...> obj` has been  compared to something - and only after it you can decide if you can get it by overloaded `operator*` or by `getNullable()`. 

The second problem - you can use `operator*` accidentally instead of  `getNullable()`. Even when you added this class you had several places like this - [here](https://github.com/ugeneunipro/ugene/blob/db15b659827f6e76533d0fd33178ad7565f0b858/src/corelibs/U2Formats/src/BAMUtils.cpp#L374), [here](https://github.com/ugeneunipro/ugene/blob/db15b659827f6e76533d0fd33178ad7565f0b858/src/corelibs/U2Formats/src/BAMUtils.cpp#L567) and [here](https://github.com/ugeneunipro/ugene/blob/db15b659827f6e76533d0fd33178ad7565f0b858/src/corelibs/U2Formats/src/BAMUtils.cpp#L670). In all these spots you use a usual `operator*` instead of `getNullable()`, but the parameter haven't been used with `operator==` or `operator!=` yet, so it causes an unfair `SAFE_POINT`. It's difficult to produce a scenario to get these lines and our tests do not cover it, so it was missed and already released, which isn't good, but not critical.

What about me - I stumbled several times on that, and, as far as I have been using a **very** large file in this issue, it was a typical case, when I'd been waiting for 4-5 hours for calculation and then had a `SAFE_POINT` on some of this pointers.

Reply:
The scenario you described demonstrates problems in your code, but not with a class: you tried to use a nullable pointer with no null checks. This is exactly why this pointer exists: to reveal such problems during testing.

The pointer has 0 effect in the production code: it is designed to behave like a raw pointer. It should add a custom behavior only in debug or testing builds.

The only "problem" I know about NP class that it can't come up with results during compile time and requires test runtime.

The rule of thumb for this class: see NP in the method: make a null check. Tired of the null checks -> convert it to a raw pointer and use the raw pointer (but you have to make the null check during the conversion).